neuron-explainer/neuron_explainer/explanations/simulator.py

```diff
@@ -590,51 +594,183 @@ def _format_record_for_logprob_free_simulation(
             activation_record.activations, max_activation=max_activation
         )
     for i, token in enumerate(activation_record.tokens):
+        # Edge Case #3: End tokens confuse the chat-based simulator. Replace end token with "not end token".
+        if token.strip() == END_OF_TEXT_TOKEN:
+            token = END_OF_TEXT_TOKEN_REPLACEMENT
         # We use a weird unicode character here to make it easier to parse the response (can split on "༗\n").
         if include_activations:
             response += f"{token}\t{normalized_activations[i]}༗\n"
         else:
             response += f"{token}\t༗\n"
     return response
 
+def _format_record_for_logprob_free_simulation_json(
+    neuron: int,
+    explanation: str,
+    activation_record: ActivationRecord,
+    include_activations: bool = False,
+    max_activation: Optional[float] = None,
+) -> str:
+    if include_activations:
+        assert max_activation is not None
+        assert len(activation_record.tokens) == len(
+            activation_record.activations
+        ), f"{len(activation_record.tokens)=}, {len(activation_record.activations)=}"
+        normalized_activations = normalize_activations(
+            activation_record.activations, max_activation=max_activation
+        )
+    return json.dumps({
+        "neuron": neuron,
+        "explanation": explanation,
+        "activations": [
+            {
+                "token": token,
+                "activation": normalized_activations[i] if include_activations else None
+            } for i, token in enumerate(activation_record.tokens)
+        ]
+    })
+
+def _parse_no_logprobs_completion_json(
+    completion: str,
+    tokens: Sequence[str],
+) -> Sequence[float]:
+    """
+    Parse a completion into a list of simulated activations. If the model did not faithfully
+    reproduce the token sequence, return a list of 0s. If the model's activation for a token
+    is not a number between 0 and 10 (inclusive), substitute 0.
+
+    Args:
+        completion: completion from the API
+        tokens: list of tokens as strings in the sequence where the neuron is being simulated
+    """
+
+    logger.debug("for tokens:\n%s", tokens)
+    logger.debug("received completion:\n%s", completion)
+
+    zero_prediction = [0] * len(tokens)
+
+    try:
+        completion = json.loads(completion)
+        if "activations" not in completion:
+            logger.error("The key 'activations' is not in the completion:\n%s", json.dumps(completion))
+            return zero_prediction
+        activations = completion["activations"]
+        if len(activations) != len(tokens):
+            logger.error("Tokens and activations length did not match:\n%s", json.dumps(completion))
+            return zero_prediction
+        predicted_activations = []
+        # check that there is a token and activation value
+        # no need to double check the token matches exactly
```

Comment:
where's the first check?

Reply:
I think I should have placed this comment one line lower. If you mean the token and activation value check:
token check is line 665 `if "token" not in activation:`
activation check is line 669 `if "activation" not in activation:`

Or do you mean a different check?

Reply:
If I'm understanding the code correctly it looks like you only check that the number of tokens is as expected, but you don't check that any of the tokens individually are correct. is that right?